# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
path = os.getcwd()
print(path)

C:\Users\j9903\Desktop\SpringBoard\Assignments\Capstone Project 1


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df0 = pd.read_csv(path + '\\data\\step3_output.csv')

In [4]:
df0.head()

Unnamed: 0                 Name    state  summit_elev  vertical_drop  \
0           0       Alyeska Resort   Alaska         3939           2500   
1           1  Eaglecrest Ski Area   Alaska         2600           1540   
2           2     Hilltop Ski Area   Alaska         2090            294   
3           3     Arizona Snowbowl  Arizona        11500           2300   
4           4  Sunrise Park Resort  Arizona        11100           1800   

   trams  fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0      1        0.0          0          2     2  ...             1610.0   
1      0        0.0          0          0     0  ...              640.0   
2      0        0.0          0          0     0  ...               30.0   
3      0        0.0          1          0     2  ...              777.0   
4      0        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         0  
1          53.0               90.0             0.0         0  
2          34.0              152.0            30.0         0  
3          89.0              122.0             0.0         1  
4          78.0              104.0            80.0         1  

[5 rows x 27 columns]

In [5]:
df0.columns

Index(['Unnamed: 0', 'Name', 'state', 'summit_elev', 'vertical_drop', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters'],
      dtype='object')

In [6]:
df0 = df0.drop('Unnamed: 0', axis=1)

In [7]:
df0.describe()

summit_elev  vertical_drop       trams   fastEight   fastSixes  \
count    330.000000     330.000000  330.000000  330.000000  330.000000   
mean    4591.818182    1215.427273    0.172727    0.003030    0.184848   
std     3735.535934     947.864557    0.559946    0.055048    0.651685   
min      315.000000      60.000000    0.000000    0.000000    0.000000   
25%     1403.750000     461.250000    0.000000    0.000000    0.000000   
50%     3127.500000     964.500000    0.000000    0.000000    0.000000   
75%     7806.000000    1800.000000    0.000000    0.000000    0.000000   
max    13487.000000    4425.000000    4.000000    1.000000    6.000000   

        fastQuads        quad     triple      double     surface  ...  \
count  330.000000  330.000000  330.00000  330.000000  330.000000  ...   
mean     1.018182    0.933333    1.50000    1.833333    2.621212  ...   
std      2.198294    1.312245    1.61913    1.815028    2.059636  ...   
min      0.000000    0.000000    0.00000    0.000000    0.000000  ...   
25%      0.000000    0.000000    0.00000    1.000000    1.000000  ...   
50%      0.000000    0.000000    1.00000    1.000000    2.000000  ...   
75%      1.000000    1.000000    2.00000    3.000000    3.000000  ...   
max     15.000000    8.000000    8.00000   14.000000   15.000000  ...   

       SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear    yearsOpen  \
count         330.000000      330.000000        330.000000   330.000000   
mean          739.801223      174.873239         97.315152    63.463636   
std          1807.868079      242.378848         52.680468   109.319671   
min             8.000000        2.000000          0.000000     0.000000   
25%            85.000000       55.000000         75.250000    50.000000   
50%           200.000000      132.500000        105.500000    58.000000   
75%           739.801223      192.000000        130.000000    68.750000   
max         26819.000000     3379.000000        305.000000  2019.000000   

       averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
count       330.000000    330.000000    330.000000         330.000000   
mean        185.316456     57.916957     64.166810         102.954545   
std         133.424094     23.898798     22.571324          50.910444   
min          18.000000     15.000000     17.000000           0.000000   
25%          71.250000     43.000000     49.000000          90.000000   
50%         158.000000     57.916957     64.166810         112.500000   
75%         278.750000     67.000000     72.000000         136.000000   
max         669.000000    179.000000    179.000000         305.000000   

       NightSkiing_ac    clusters  
count      330.000000  330.000000  
mean        56.890909    0.357576  
std         93.464882    0.486305  
min          0.000000    0.000000  
25%          0.000000    0.000000  
50%         27.000000    0.000000  
75%         83.750000    1.000000  
max        650.000000    2.000000  

[8 rows x 24 columns]

In [8]:
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 330 entries, 0 to 329
Data columns (total 26 columns):
Name                 330 non-null object
state                330 non-null object
summit_elev          330 non-null int64
vertical_drop        330 non-null int64
trams                330 non-null int64
fastEight            330 non-null float64
fastSixes            330 non-null int64
fastQuads            330 non-null int64
quad                 330 non-null int64
triple               330 non-null int64
double               330 non-null int64
surface              330 non-null int64
total_chairs         330 non-null int64
Runs                 330 non-null float64
TerrainParks         330 non-null float64
LongestRun_mi        330 non-null float64
SkiableTerrain_ac    330 non-null float64
Snow Making_ac       330 non-null float64
daysOpenLastYear     330 non-null float64
yearsOpen            330 non-null float64
averageSnowfall      330 non-null float64
AdultWeekday         330 non-null f

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [9]:
df = pd.concat([df0.drop('state', axis=1), pd.get_dummies(df0["state"])], axis=1)

In [31]:
df.shape

(330, 60)

In [10]:
df.head(10)

Name  summit_elev  vertical_drop  trams  \
0                 Alyeska Resort         3939           2500      1   
1            Eaglecrest Ski Area         2600           1540      0   
2               Hilltop Ski Area         2090            294      0   
3               Arizona Snowbowl        11500           2300      0   
4            Sunrise Park Resort        11100           1800      0   
5  Yosemite Ski & Snowboard Area         7800            600      0   
6                  Bear Mountain         8805           1665      0   
7                    Bear Valley         8500           1900      0   
8         Boreal Mountain Resort         7700            500      0   
9                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          2     2       0       0  ...             0   
1        0.0          0          0     0       0       4  ...             0   
2        0.0          0          0     0       1       0  ...             0   
3        0.0          1          0     2       2       1  ...             0   
4        0.0          0          1     2       3       1  ...             0   
5        0.0          0          0     0       1       3  ...             0   
6        0.0          0          2     1       2       3  ...             0   
7        0.0          1          1     0       2       4  ...             0   
8        0.0          0          1     1       3       1  ...             0   
9        0.0          0          0     1       2       5  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   
5             0          0     0        0         0           0   
6             0          0     0        0         0           0   
7             0          0     0        0         0           0   
8             0          0     0        0         0           0   
9             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  
5              0          0        0  
6              0          0        0  
7              0          0        0  
8              0          0        0  
9              0          0        0  

[10 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [11]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X) 

In [12]:
X.shape

(330, 58)

In [13]:
X_scaled.shape

(330, 58)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [14]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [15]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [16]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [17]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
explained_variance_score(y_test,y_pred)

0.9404111811904321

In [18]:
mean_absolute_error(y_test,y_pred)

4.892241477165638

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [19]:
lm.intercept_

64.09330396245056

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [20]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Coefficient
total_chairs       2.427756e+13
fastQuads          9.203676e+12
surface            8.623154e+12
double             7.599043e+12
triple             6.778869e+12
quad               5.494022e+12
New York           3.843861e+12
Michigan           3.627561e+12
Colorado           3.196092e+12
California         3.127673e+12
Pennsylvania       2.984623e+12
Wisconsin          2.752056e+12
New Hampshire      2.752056e+12
fastSixes          2.728435e+12
Vermont            2.668907e+12
Minnesota          2.582498e+12
Utah               2.492492e+12
Montana            2.398483e+12
Idaho              2.398483e+12
trams              2.344347e+12
Massachusetts      2.299981e+12
Oregon             2.196380e+12
Washington         2.196380e+12
New Mexico         2.086922e+12
Maine              2.086922e+12
Wyoming            1.970632e+12
North Carolina     1.711909e+12
Ohio               1.565162e+12
Connecticut        1.565162e+12
Nevada             1.402075e+12
Illinois           1.402075e+12
Virginia           1.402075e+12
West Virginia      1.402075e+12
Alaska             1.216094e+12
Iowa               1.216094e+12
South Dakota       9.944534e+11
Indiana            9.944534e+11
Arizona            9.944534e+11
Missouri           9.944534e+11
New Jersey         9.944534e+11
Rhode Island       7.042558e+11
Maryland           7.042558e+11
Tennessee          7.042558e+11
fastEight          2.304723e+11
AdultWeekday       1.934131e+01
summit_elev        1.816450e+00
Runs               1.709656e+00
vertical_drop      9.841541e-01
NightSkiing_ac     7.480469e-01
clusters           6.202393e-01
daysOpenLastYear   6.003418e-01
LongestRun_mi      4.511719e-01
yearsOpen          4.487305e-01
projectedDaysOpen  4.167480e-01
Snow Making_ac     3.725586e-01
averageSnowfall    1.711426e-01
SkiableTerrain_ac  1.528320e-01
TerrainParks       2.026367e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [21]:
df_m2 = df0.copy()

In [22]:
X_m2 = df_m2.drop(['Name','state','AdultWeekend'], axis=1)
y_m2 = df_m2.AdultWeekend
scaler_m2 = preprocessing.StandardScaler().fit(X_m2)
X_scaled_m2 =scaler_m2.transform(X_m2)
y_m2 = y_m2.ravel()
X_train_m2, X_test_m2, y_train_m2, y_test_m2 = train_test_split(X_scaled_m2, y_m2, test_size=0.25, random_state=1)

In [23]:
lm_m2 = linear_model.LinearRegression()
model_m2 = lm_m2.fit(X_train_m2,y_train_m2)
y_pred_m2 = model_m2.predict(X_test_m2)

In [24]:
print("explained_variance_score = " + str(explained_variance_score(y_test_m2,y_pred_m2)))
print("mean_absolute_error = " + str(mean_absolute_error(y_test_m2,y_pred_m2)))
print("Intercept of the model = " + str(lm_m2.intercept_))

explained_variance_score = 0.9334949124957841
mean_absolute_error = 5.091002000817155
Intercept of the model = 64.04575282803901


In [25]:
pd.DataFrame(abs(lm_m2.coef_), X_m2.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Coefficient
AdultWeekday         19.891878
clusters              5.928500
summit_elev           4.630346
Runs                  1.993381
triple                1.439498
surface               1.370549
averageSnowfall       1.240615
fastQuads             1.095524
vertical_drop         1.062243
quad                  1.059678
NightSkiing_ac        0.761536
projectedDaysOpen     0.715232
total_chairs          0.662522
LongestRun_mi         0.443193
SkiableTerrain_ac     0.403323
fastSixes             0.398569
yearsOpen             0.357766
fastEight             0.220547
double                0.166126
TerrainParks          0.143809
Snow Making_ac        0.090747
trams                 0.068567
daysOpenLastYear      0.055025

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [26]:
df_m3 = df0.copy()

In [27]:
X_m3 = df_m3.drop(['Name','state','AdultWeekend','summit_elev'], axis=1)
y_m3 = df_m3.AdultWeekend
scaler_m3 = preprocessing.StandardScaler().fit(X_m3)
X_scaled_m3 =scaler_m3.transform(X_m3)
y_m3 = y_m3.ravel()
X_train_m3, X_test_m3, y_train_m3, y_test_m3 = train_test_split(X_scaled_m3, y_m3, test_size=0.25, random_state=1)

In [28]:
lm_m3 = linear_model.LinearRegression()
model_m3 = lm_m3.fit(X_train_m3,y_train_m3)
y_pred_m3 = model_m3.predict(X_test_m3)

In [29]:
print("explained_variance_score = " + str(explained_variance_score(y_test_m3,y_pred_m3)))
print("mean_absolute_error = " + str(mean_absolute_error(y_test_m3,y_pred_m3)))
print("Intercept of the model = " + str(lm_m3.intercept_))

explained_variance_score = 0.9326122385106692
mean_absolute_error = 5.128384719184385
Intercept of the model = 64.02834587024049


In [30]:
pd.DataFrame(abs(lm_m3.coef_), X_m3.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Coefficient
AdultWeekday         20.093854
clusters              2.621411
vertical_drop         2.102951
Runs                  1.515356
triple                1.425470
surface               1.141511
quad                  1.118070
averageSnowfall       1.056774
fastQuads             0.859719
projectedDaysOpen     0.769349
total_chairs          0.659297
fastSixes             0.515291
TerrainParks          0.481739
NightSkiing_ac        0.417335
LongestRun_mi         0.390924
yearsOpen             0.336765
Snow Making_ac        0.324899
trams                 0.265079
daysOpenLastYear      0.197553
double                0.174753
SkiableTerrain_ac     0.131614
fastEight             0.081277

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9404111811904321 | 4.892241477165638 |'base_elev'|
| Model 2. | 0.9334949124957841 | 5.091002000817155 |'state','base_elev'|
| Model 3. | 0.9326122385106692 | 5.128384719184385 |'state','summit_elev','base_elev'|

Model Selection: Model 1

The success criteria of the model performance is 

1. Smaller MAE (ideally <= 40; small MAE suggests the model is great at prediction, while a large MAE suggests that your model may have trouble in certain areas. A MAE of 0 means that your model is a perfect predictor of the outputs but this will almost never happen)
2. Higher EV (for EV, the best possible score is 1.0 and lower values are worse)

Therefore model 2 and 3 are not selected because their MAE values are higher and EV values are lower than model 1.